File: string/nativesearch/string_search.c
/**
 * Copyright © https://github.com/jarry All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <memory.h>

int find(char* str, char* text)
{
  int text_len = strlen(text);
  int str_len = strlen(str);
  // 两个循环，外层是被查找文本，内循环是查找字符串
  for (int i = 0; i < text_len; i++)
  {
    int j = 0;
    for (; j < str_len; j++)
    {
      // 当遇到不有不相等时，跳出从文本下一个字符开始比较
      if (str[j] != text[i + j])
      {
        break;
      }
    }
    // 如果查找字符串全部比较完成表示成功匹配
    if (j == str_len)
    {
      return i;
    }
  }

  // 如果文本全部比较完还是没有查找到，则返回-1
  return -1;
}

int find2(char* str, char* text)
{
  int text_len = strlen(text);
  int str_len = strlen(str);
  // 两个循环，外层是被查找文本，内循环是查找字符串
  for (int i = 0, l = text_len - str_len + 1; i < l; i++)
  {
    int j = 0;
    for (; j < str_len; j++)
    {
      // 循环长度优化+剩余字符优化
      // 如果被查找的长度比要查找的还少，则可以返回-1
      if (text_len - i < str_len - j)
      {
        // console.log('less length', i, j)
        return -1;
      }
      // 当遇到不有不相等时，跳出从文本下一个字符开始比较
      if (str[j] != text[i + j])
      {
        break;
      }
    }
    // 如果查找字符串全部比较完成表示成功匹配
    if (j == str_len)
    {
      return i;
    }
  }

  // 如果文本全部比较完还是没有查找到，则返回-1
  return -1;
}

int main()
{
  /* find test */
  float startTime = clock();
  printf("find start");
  char str1[] = "ABC";
  char text1[] = "AABABC";
  int result = find(str1, text1);       // 3
  printf("\nfind('ABC', 'AABABC'), %d", result);

  char str2[] = "AAB";
  char text2[] = "AAAABC";
  result = find(str2, text2);;          // 2
  printf("\nfind('AAB', 'AAAABC'), %d", result);

  char str3[] = "ABC";
  char text3[] = "AABAC";
  result = find(str3, text3);;          // -1
  printf("\nfind('ABC', 'AABAC'), %d", result);
  printf("\ntime: %f ms.", ((clock() - startTime) / CLOCKS_PER_SEC * 1000));
  printf("\nfind end\n");

  /* find2 test */

  startTime = clock();
  printf("find2 start");
  result = find2(str1, text1);           // 3
  printf("\nfind('ABC', 'AABABC'), %d", result);

  result = find2(str2, text2);;          // 2
  printf("\nfind('AAB', 'AAAABC'), %d", result);

  result = find2(str3, text3);;          // -1
  printf("\nfind('ABC', 'AABAC'), %d", result);
  printf("\ntime: %f ms.", ((clock() - startTime) / CLOCKS_PER_SEC * 1000));
  printf("\nfind2 end\n");
}

/**
jarry@jarrys-MacBook-Pro nativesearch % gcc string_search.c -o string_search
jarry@jarrys-MacBook-Pro nativesearch % ./string_search                     
find start
find('ABC', 'AABABC'), 3
find('AAB', 'AAAABC'), 2
find('ABC', 'AABAC'), -1
time: 0.041000 ms.
find end
find2 start
find('ABC', 'AABABC'), 3
find('AAB', 'AAAABC'), 2
find('ABC', 'AABAC'), -1
time: 0.006000 ms.
find2 end
*/

File: string/nativesearch/string_search.dart

/**
 * @param pattern 查找串
 * @text 目标内容
 */
int find(String pattern, String text) {
  for (int i = 0; i < text.length; i++) {
    int j;
    for (j = 0; j < pattern.length; j++) {
      if (text[i + j] != pattern[j]) {
        break;
      }
    }
    if (j == pattern.length) {
      return i;
    }
  }
  return -1;
}

int find1(String pattern, String text) {
  int i = 0;
  int j = 0;
  while (j < pattern.length && i < text.length) {
    if (pattern[j] != text[i]) {
      i = i - j + 1;
      j = 0;
    } else {
      i++;
      j++;
    }
    if (j == pattern.length) {
      return i - j;
    }
  }
  return -1;
}

int find2(String pattern, String text) {
  int i = 0;
  int j = 0;
  print("find2");
  for (; j < pattern.length;) {
    for (; i < text.length;) {
      if (pattern[j] != text[i]) {
        i = i - j + 1;
        j = 0;
      } else {
        i++;
        j++;
      }
      break;
    }
    if (pattern.length - j > text.length - i) {
      break;
    }
    if (j == pattern.length) {
      return i - j;
    }
  }
  return -1;
}

void main() {
  print("Hello, search start：");
  int result = -1;
  
  // find() test
  result = find("ABC", "AABABC");
  print("find ABC in AABABC. the index is 3: $result");  // 3
  result = find("AABC", "AAAABC");
  print("find `AABC` in `AAAABC` the index is 2: $result"); // 2
  result = find("ABC", "AABAC");
  print("find `ABC` in `AABAC` the index is -1: $result"); // -1

  // find1() test
  result = find1("ABC", "AABABC");
  print("find1 ABC in AABABC. the index is 3: $result");  // 3
  result = find1("AABC", "AAAABC");
  print("find1 `AABC` in `AAAABC` the index is 2: $result"); // 2
  result = find1("ABC", "AABAC");
  print("find1 `ABC` in `AABAC` the index is -1: $result"); // -


  // find2() test
  result = find2("ABC", "AABABC");
  print("find2 ABC in AABABC. the index is 3: $result");  // 3
  result = find2("AABC", "AAAABC");
  print("find2 `AABC` in `AAAABC` the index is 2: $result"); // 2
  result = find2("ABC", "AABAC");
  print("find2 `ABC` in `AABAC` the index is -1: $result"); // -1
  result = find2("ABBCD", "ABBCAABCAABBCDCD");
  print("find2 `ABC` in `AABAC` the index is 9: $result"); // 9
}

/* Test
jarry@jarrys-MacBook-Pro nativesearch % dart --version
Dart SDK version: 2.14.1 (stable) (Wed Sep 8 13:33:08 2021 +0200) on "macos_x64"
jarry@jarrys-MacBook-Pro nativesearch % dart string_search.dart 
*/